src/subcommand/torrent/announce.rs

```diff
@@ -0,0 +1,135 @@
+use crate::common::*;
+
+const URI_HELP: &str = "Announce an infohash and list the response (a compact peer list).";
+
+const INPUT_HELP: &str =
+  "Generate a compact peer list from a metainfo at `INPUT`. If `INPUT` is `-`, read \
+                          metainfo from standard input.";
+
+const INPUT_FLAG: &str = "input-flag";
+
+const INPUT_POSITIONAL: &str = "<INPUT>";
+
+#[derive(StructOpt)]
+#[structopt(
+  help_message(consts::HELP_MESSAGE),
+  version_message(consts::VERSION_MESSAGE),
+  about(URI_HELP)
+)]
+pub(crate) struct Announce {
+  #[structopt(
+    name = INPUT_FLAG,
+    long = "input",
+    short = "i",
+    value_name = "INPUT",
+    empty_values(false),
+    parse(try_from_os_str = InputTarget::try_from_os_str),
+    help = INPUT_HELP,
+  )]
+  input_flag: Option<InputTarget>,
+  #[structopt(
+    name = INPUT_POSITIONAL,
+    value_name = "INPUT",
+    empty_values(false),
+    parse(try_from_os_str = InputTarget::try_from_os_str),
+    required_unless = INPUT_FLAG,
+    conflicts_with = INPUT_FLAG,
+    help = INPUT_HELP,
+  )]
+  input_positional: Option<InputTarget>,
+}
+
+impl Announce {
+  pub(crate) fn run(self, env: &mut Env, options: &Options) -> Result<(), Error> {
+    let target = xor_args(
+      "input_flag",
+      &self.input_flag,
+      "input_positional",
+      &self.input_positional,
+    )?;
+    let input = env.read(target)?;
+
+    let infohash = Infohash::from_input(&input)?;
+    let metainfo = Metainfo::from_input(&input)?;
+
+    let mut rng = rand::thread_rng();
+    let peer_id: [u8; 20] = rng.gen();
+    let mut peer_list = Vec::new();
+
+    if metainfo.trackers().peekable().peek().is_none() {
+      if !options.quiet {
+        println!("Supplied metainfo specifies no trackers.");
+      }
+      return Err(Error::NoPeerSource);
+    }
+
+    if !options.quiet {
+      println!("[1/2] Announcing {} to trackers.", &infohash);
```

Comment:
All printing to stderr and stdout should be done via the `env` object, so it can be captured when running tests. There are macros `err`, `errln`, `out`, and `outln` that make doing this convenient.

Reply:
Any downside to adding something like this to avoid !quiet pattern?
```rust
  ($env:expr, $opt:expr, $fmt:expr, $($arg:tt)*) => {
    if !opt.quiet {
      write!($env.out_mut(), $fmt, $($arg)*).context(crate::error::Stdout)
    }
  };
```

Reply:
Good idea! Actually, I think it might be better to just suppress all output to stderr if the `--quiet` flag is passed. This can be done centrally in `OutputStream`. I just opened #473 to track this. You can skip adding `if !quiet`, since this is simple enough that I should be able to add it soon.

Reply:
This is done: #474.